app/services/organizations/task_service.rb

```diff
@@ -0,0 +1,38 @@
+class Organizations::TaskService
+  def initialize(task)
+    @task = task
+  end
+
+  def create_next
+    if need_next?
+      Task.create(
+        pet_id: @task.pet_id,
+        name: @task.name,
+        description: @task.description,
+        recurring: true,
+        next_due_date_in_days: @task.next_due_date_in_days,
+        due_date: next_due_date
+      )
+    end
+  end
+
+  private
+
+  def next_due_date
+    return nil unless @task.due_date
+
+    if overdue?
+      @task.updated_at + @task.next_due_date_in_days.days
+    else
+      @task.due_date + @task.next_due_date_in_days.days
+    end
+  end
+
+  def need_next?
+    @task.recurring && @task.completed_previously_changed?(from: false, to: true)
+  end
+
+  def overdue?
```

Comment:
We now have a method on Task `def overdue?` to do this. It uses `Time.current` which I think should work fine.

One change it could have, is to check if the task is complete or not...do we want to consider a completed task overdue? Probably not.

Reply:
I think for a more general `overdue?` method (one that needs to work outside of the task service object) it should be `Time.current` and not `updated_at`. I was thinking when the service object checks if a task is overdue it only does it if the update was a change from incomplete to complete, meaning the `updated_at` value was the time at which the task became complete. But I don't think `updated_at` would be the right comparison in the general case. Like you said, there's no guarantee in other cases that the `updated_at` value would reflect the time at which the task became completed since there could be other updates that had been made. 

I agree that probably a task that is complete shouldn't be considered overdue. But if `overdue?` becomes something like: 

```
def overdue? 
  due_date < Time.current if due_date && !completed
end 
```

and the service object uses it, won't tasks always be not overdue when the service object gets instantiated? (since those tasks will be completed). 

One option, could be to split off the time aspect into its own method like: 

```
def due_date_passed?
 due_date < Time.current if due_date
end
```

and then overdue? could be:

```
def overdue?
  due_date_passed? && !completed
end
```

Then the service object would use `due_date_passed?` and the view that is checking for overdue-ness could remove the conjunction. 

I don't know if there will be other things that want to check for overdue-ness but if so it might be worth making it more accurate? Intuitive? -- I don't know the word I'm looking for here...

Reply:
Ah yes good points, you are right! Then I think your original approach makes sense. We can have a method on the Task for general use, and have a method on the Service object that does what it needs to do.